packages/controller-utils/src/create-service-policy.ts

```diff
@@ -0,0 +1,181 @@
+ * designated time period has passed; this particular behavior is primarily
+ * designed for services that wrap API calls so as not to make needless HTTP
+ * requests when the API is down and to be able to recover when the API comes
+ * back up. In addition, hooks allow for responding to certain events, one of
+ * which can be used to detect when an HTTP request is performing slowly.
+ *
+ * Internally, the executor object makes use of the retry and circuit breaker
+ * policies from the [Cockatiel](https://www.npmjs.com/package/cockatiel)
+ * library; see there for more.
+ *
+ * @param options - The options to this function.
+ * @param options.maxRetries - The maximum number of times that a failing action
+ * should be re-run before giving up. Defaults to 3.
+ * @param options.maxConsecutiveFailures - The maximum number of times that
+ * the action is allowed to fail before pausing further retries. Defaults to 12.
+ * @param options.circuitBreakDuration - The length of time (in milliseconds) to
+ * pause retries of the action after the number of failures reaches
+ * `maxConsecutiveFailures`.
+ * @param options.degradedThreshold - The length of time (in milliseconds) that
+ * governs when an action is regarded as degraded (affecting when `onDegraded`
+ * is called). Defaults to 5 seconds.
+ * @param options.onBreak - A function which is called when the action fails too
+ * many times in a row.
+ * @param options.onDegraded - A function which is called when the action
+ * succeeds before `maxConsecutiveFailures` is reached, but takes more time
+ * than the `degradedThreshold` to run.
+ * @returns A Cockatiel policy object that can be used to run an arbitrary
+ * action (a function).
+ * @example
+ * To use the policy, call `execute` on it and pass a function:
+ * ``` ts
+ * const policy = createServicePolicy({
+ *   maxRetries: 3,
+ *   maxConsecutiveFailures: 3,
+ *   circuitBreakDuration: 5000,
+ *   degradedThreshold: 2000,
+ *   onBreak: () => {
+ *     console.log('Circuit broke');
+ *   },
+ *   onDegraded: () => {
+ *     console.log('Service is degraded');
+ *   }
+ * });
+ *
+ * await policy.execute(async () => {
+ *   const response = await fetch('https://some/url');
+ *   return await response.json();
+ * });
+ * ```
+ * You may wish to store `policy` in a single place and reuse it each time you
+ * want to execute your action. For instance:
+ * ``` ts
+ * class Service {
+ *   constructor() {
+ *     this.#policy = createServicePolicy({
+ *       maxRetries: 3,
+ *       maxConsecutiveFailures: 3,
+ *       circuitBreakDuration: 5000,
+ *       degradedThreshold: 2000,
+ *     });
+ *   }
+ *
+ *   async fetch() {
+ *     return await this.#policy.execute(async () => {
+ *       const response = await fetch('https://some/url');
+ *       return await response.json();
+ *     });
+ *   }
+ * }
+ * ```
+ */
+export function createServicePolicy({
+  maxRetries = DEFAULT_MAX_RETRIES,
```

Comment:
Should this be called `retries` instead of `maxRetries`? Alternatively, should this be called `maxAttempts` instead of `maxRetries` (in which case the user would have to specify 1 more than the number of retries they expect)?